src/base/locked_queue.h

```diff
@@ -61,56 +67,112 @@ class LockedQueue {
 
     return true;
   }
 
   // Returns disengaged object only when this queue is closed and empty.
-  Optional Pop(ui32 shard = DEFAULT_SHARD) THREAD_SAFE {
+  Optional Pop() THREAD_SAFE {
     UniqueLock lock(pop_mutex_);
-    pop_condition_.wait(lock, [this] { return closed_ || !queue_.empty(); });
+    index_.WaitDefaultShard(lock,
+                            [this] { return closed_ || !queue_.empty(); });
     if (closed_ && queue_.empty()) {
       return Optional();
     }
 
-    auto it = index_.Get(shard, queue_.begin());
-    Optional&& obj = std::move(*it);
-    queue_.erase(it);
-    --size_;
-    return std::move(obj);
+    return RemoveTaskFromQueue(
+        index_.GetWithHint(DEFAULT_SHARD, queue_.begin()));
   }
 
   // Returns disengaged object only when this queue is closed and empty, or pool
   // is shutting down.
-  Optional Pop(const WorkerPool& pool, ui32 shard = DEFAULT_SHARD) THREAD_SAFE {
+  Optional Pop(const WorkerPool& pool,
+               const ui32 shard_queue_limit = NOT_STRICT_SHARDING,
+               const ui32 shard = DEFAULT_SHARD) THREAD_SAFE {
     DCHECK(timeout_ > Seconds::zero());
 
     UniqueLock lock(pop_mutex_);
-    do {
-      pop_condition_.wait_for(lock, timeout_);
-    } while (!closed_ && queue_.empty() && !pool.IsShuttingDown());
+    if (shard_queue_limit == NOT_STRICT_SHARDING) {
+      return PopWithHint(pool, lock, shard);
+    } else {
+      return PopStrict(pool, lock, shard_queue_limit, shard);
+    }
+  }
+
+ private:
+  Optional PopWithHint(const WorkerPool& pool, UniqueLock& lock,
+                       const ui32 shard) THREAD_UNSAFE {
+    // One can't wait for condition using predicate and timed wait here as once
+    // the waiting timed out, the |pool.IsShuttingDown()| should be checked and
+    // wait again if pool doesn't shutting down.
+    while (!closed_ && queue_.empty() && !pool.IsShuttingDown()) {
+      index_.WaitShardFor(shard, lock, timeout_);
+    }
     if ((closed_ && queue_.empty()) || pool.IsShuttingDown()) {
       return Optional();
     }
+    return RemoveTaskFromQueue(index_.GetWithHint(shard, queue_.begin()));
+  }
+
+  Optional PopStrict(const WorkerPool& pool, UniqueLock& lock,
+                     const ui32 shard_queue_limit,
+                     const ui32 shard) THREAD_UNSAFE {
+    DCHECK(shard_queue_limit != NOT_STRICT_SHARDING);
+    // Check if there's a shard with number of tasks that exceed queue limit
+    // AND current shard has no tasks to do. If so, steal a task from overloaded
+    // shard.
+    const ui32 maybe_overloaded_shard =
+        index_.MaybeOverloadedShard(shard_queue_limit, shard);
+    if (shard != maybe_overloaded_shard) {
+      return RemoveTaskFromQueue(index_.GetStrict(maybe_overloaded_shard));
+    }
 
-    auto it = index_.Get(shard, queue_.begin());
-    Optional&& obj = std::move(*it);
-    queue_.erase(it);
+    auto shard_is_closed = [&] {
+      return closed_ || shard_queue_limit == GET_SHARD_DOWN;
```

Comment:
Please, add comment with scenario for `shard_queue_limit` - I spent 15 minutes to understand it.